Do not assign option.Config.Opts after init

jrajahalme opened the pull request:
Access to `option.Config.Opts` is not protected against the pointer changing after init. Fix by removing pointer change on config change revert.

Clean up `ConfigPatchMutex` use by removing unused setters and unnecessary getters.

Now that `ConfigPatchMutex` is only used to guard patching of `Opts` by the daemon, we can remove `ConfigPatchMutex` altogether, as daemon config events are already serialized via an event queue.
This simplifies testing `option.DaemonConfig` by removing the need to initialize the mutex pointer.

Finally, unexport `IntOptions` fields (`option.DaemonConfig.Opts`) to protect against accidental bypass of the mutex within.

jrajahalme committed:
option: Remove [GSet]IPv[46]NativeRoutingCIDR accessors

Native routing CIDR options are only set during agent bootstrap before
options are exposed, and never changed later. Simplify the locking around
options.Config by removing these accessors and accessing the fields
directly.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
daemon: Do not replace option.Config.Opts pointer

option.Config.Opts pointer change on revert was protected by a mutex, but
access to option.Config.Opts pointer never took the mutex. This would
allow a caller to take the internal Opts mutex, and then release one from
another instance, which would panic.

Fix by never replacing the Opts pointer after initialization. Keep it as
a pointer to be able to take a shallow copy of option.Config for
checksumming the invariable parts of the config.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

derailed reviewed (approved):
@jrajahalme Nice update. Thank you Jarno!

jrajahalme committed:
daemon: Remove option.DaemonConfig.ConfigPatchMutex

option.DaemonConfig.ConfigPatchMutex is no longer needed, since
option.DaemonConfig.Opts is already protected against concurrent patching
via an event queue that serializes config patch API calls.

Note that reading the non-mutable parts of the DaemonConfig was not
protected by the ConfigPatchMutex anyway, and DaemonConfig.Opts has it's
own internal mutex for protecting against concurrent access during the
config patch operation.

Removing ConfigPatchMutex simplifies testing code as we can get rid of
the mutex pointer initialization.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
endpointmanager: Remove unnecessary SetDefaultConfiguration call

Endpoint's default configuration options are already set earlier in the
restore process at Daemon.restoreOldEndpoints().

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
option: Unexport all fields in IntOptions

Guard against unprotected access by making IntOptions fields unexported.
Take the internal mutex while encoding JSON.

Manually validated that JSON encoding remains the same before and after
this change.

Initialize Options with the endpoint so that there is no need to check
for that later and that the IntOptions pointer need not change after
init.

Fix comment on endpoint option validation. Endpoint.SetDefaultOpts does
not validate the given 'opts', but sets each option in
option.EndpointOptionsLibrary to the value in 'opts', as well as sets the
debug option value.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
endpoint: Fix option validation on restart

The SetDefaultOpts() call after endpoint unmarshal did not really
validate anything, it simply re-set the same option values again using
the SetValidated() call, ranging over the options library, and leaving
options not in the library in the map.

Change the JSON unmarshal to silently discard unsupported options
instead, and pass a nil map to SetDeafaultOpts(). Note that Endpoint's
options are later set to the agent defaults in restoreOldEndpoints() via
SetDefaultConfiguration() unless option.Config.KeepConfig is set.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
option: Use library receiver

Use library as a receiver instead of passing it as a (pointer) argument.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

gandro reviewed (approved):
Looks good for my codeowners. The native routing CIDR can be changed rather late in the boostrap (search for autoDetectIPv4NativeRoutingCIDR), but it's still before `newDaemon` returns and the derived native routing CIDR is not stored in the DaemonConfig, so the change for that is fine!

tommyp1ckles commented on pkg/endpoint/endpoint.go:
```diff
@@ -904,8 +904,9 @@ func parseEndpoint(ctx context.Context, owner regeneration.Owner, policyGetter p
-	// Validate the options that were parsed
-	ep.SetDefaultOpts(ep.Options)
+	// Set default options, unsupported options were already dropped by
+	// ep.Options.UnmarshalJSON
```
think this should be `ep.UnmarshalJSON`,  ep.Options doesn't appear to have a UnmarshalJSON func.

jrajahalme replied:
It's here:  https://github.com/cilium/cilium/pull/33164/files#diff-87bc5ccc7580502f71b247772a6e53aec7835f895cfaf90735895630fae75b5eR177-R195

tommyp1ckles reviewed (approved):
Approving for sig-metrics

derailed reviewed (approved):
@jrajahalme LGTM

doniacld commented on pkg/endpoint/endpoint.go:
```diff
@@ -909,8 +909,9 @@ func parseEndpoint(ctx context.Context, owner regeneration.Owner, policyGetter p
-	ep.SetDefaultOpts(ep.Options)
+	// Set default options, unsupported options were already dropped by
+	// ep.Options.UnmarshalJSON
+	ep.SetDefaultOpts(nil)
```
nit: typo in the commit message `SetDeafaultOpts()` --> `SetDefaultOpts()`

jrajahalme replied:
I'll let this slide :-)

doniacld reviewed (approved):
Nice work!
